### PR TITLE
0.91 lock mechanism

### LINKED ISSUE #54: Add a lock mecanism for tickets
Add a mecanism to prevent several people to work on the same ticket.

### PR DESCRIPTION
**Pull request by tomolimo**
Lock Mechanism implementation (fixes #54):

Added a new right: UNLOCK in define.php
Changed back modifications on Html::convDate() as it is needed! And added a new $format=null parameter to Html::convDateTime() to allow format of datetime.
Deleted a number (4204) that was in javascript block in Html::initEditorSystem() (line 3873)
Added case "0.90.1" in cliupdate.php to be able to perform update to 0.91 within console
Commented out a new Followup in ticket.class.php (used no where)

### CODE REVIEW

**Comment by orthagh** [comment]
I started my tests, one big problem :  the flag lock_use_lock_item in CFG_GLPI is not initialized (missing in db.glpi_configs so i have many notices at screen (and in php-errors.log)

File: inc/objectlock.class.php [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+ *
+ * Item locks are used to lock items like Ticket, Computer, Reminder, ..., see list in ObjectLock::$lockableObjects
+ * 
+ * @author Olivier Moron
+ * @since version 0.91
+ *
+ */
+class ObjectLock extends CommonDBTM {
+
+   static private $lockableObjects = array(
```
Comment (orthagh) [inline inc/objectlock.class.php]: This static list should be defined in config/define.php

**Comment by orthagh** [comment]
i don't like the style of dialog : 
image

It should be more neutral for a correct integration in all palette.
We can test with a silver or white background

**Comment by orthagh** [comment]
I remember the old plugin add an overlay or set in Read right the current profile to permit view of item.

when an item is locked, i have this behavior : 
image

Is it normal ?

**Comment by tomolimo** [comment]
For the style: i used the same than the one used for the "message after redirection".
For the "you don't have the rigths": no it's not normal, what did you try?

**Comment by orthagh** [comment]
I adde missing flag in my glpi_config (lock_use_lock_item)

I added all lock rights i found in profile to my current profile (super-admin, and yes you should enable by default for this profile :) )

add all itemtype in general config

logout/login

in 2 separate session (cdamas and me), i opened same object

**Comment by orthagh** [comment]
lock_item_list flag is also missing by default in glpi_configs table

**Comment by tomolimo** [comment]
I don't understand how you may have such troubles...
I checked for the lock_use_lock_item: it is missing: I'll add it
And for lock_item_list: it is initialized when you set items to lock in config form
I made also tests with two different sessions, with one in 'private' mode (otherwise your are really using the same php session id)...Or you may use two different browsers.

**Comment by tomolimo** [comment]
Dear Orthagh,
1) Could you tell me exactly what you did to get the 'vous n'avez pas les droits requis...': as I can't reproduce this behavior?
2) For the palette of the lock message: I don't use any color schema, but rather I'm using a class already defined in GLPI:  "dialogClass: 'message_after_redirect'" like in the function Html::displayMessageAfterRedirect(). I really made a copy/paste of this function :)
Thank you,
Tomolimo

**Comment by tomolimo** [comment]
Doing more tests, I saw the following functional problem:
When a user has only rights to view an item (for example: read-only on PC), then the PC will be locked. But in this case the item should be not locked as the viewing user has only read rights with his/her current profile.
I'll check deeper this and fix it.

**Comment by tomolimo** [comment]
I've prepared a new version, it fixes:
- the functional problem mentioned above,
- the missing config parameters (lock_use_lock_item and lock_item_list)
- the move of the lockable list of objects (moved to define.php)
- added UNLOCK rights to Super-Admin profile to be set by default.

Before doing a commit, I'm going to retest a migration and a fresh install.

**Comment by tomolimo** [comment]
and also fixed a problem on a non-existing rights in current profile ('global_validation')

**Comment by tomolimo** [comment]
also fixed a bug on template use at creation of new object. Templates were locked, preventing creation a another object at the same time.

**Comment by tomolimo** [comment]
Hello again,
I made a new commit with the fixes announced above.
Regarding the tests done in following configs
1) from scratch: OK
2) from our current running configuration: OK (done with cliupdate)

Regarding the tests: if you are using the same browser to simulate two different persons: you must use the 'inPrivate' feature to guarantee that browser sessions (and PHP sessions) will be different from one person to the other.

Regards,
Tomolimo

**Comment by orthagh** [comment]
ok
In case of new install, i succeded to test the feature

**Comment by orthagh** [comment]
however, it doesn't seem to work with ticket (computer ok)

**Comment by tomolimo** [comment]
Do yo uhave any error messages?

**Comment by orthagh** [comment]
no, but i have a strange behavior : 

* I open a ticket
* i duplicate the google chrome tab, no lock showing
* i switch to first, refresh, not much better
* i switch to second, refresh, lock set

**Comment by tomolimo** [comment]
the problem is HTTP 304: no reload just use the local cache...

**Comment by tomolimo** [comment]
You may check with spying the network traffic, no reload, juste local or remote cache

**Comment by orthagh** [comment]
ok. seems ok so.

I tested with google chrome private session, it's ok

**Comment by tomolimo** [comment]
yes, of course you need to be private otherwise the browser is using the same session

**Comment by orthagh** [comment]
a question :  is there a page to view list of lock (and manage them)  ?

**Comment by tomolimo** [comment]
Not for all locks, but type by type, you may use the following fields in the columns
glpi - computers - mozilla firefox_2016-01-11_16-38-02

**Comment by tomolimo** [comment]
and the unlock is via massiveactions

**Comment by orthagh** [comment]
ok fine.

My tests continue on update (with current 0.90/bugfixes db data).
I'm triggering some errors on ticket : 

```
2016-01-11 17:08:21 [27@adelaunay-ThinkPad-Edge-E320]
  *** PHP Notice(8): Undefined index: interface
  Backtrace :
  inc/ticket.class.php:218                           
  inc/ticket.class.php:578                           Ticket::canUpdate()
  inc/commonglpi.class.php:208                       Ticket->getTabNameForItem()
  inc/ticket.class.php:660                           CommonGLPI->addStandardTab()
  inc/commonglpi.class.php:158                       Ticket->defineTabs()
  inc/commonglpi.class.php:602                       CommonGLPI->defineAllTabs()
  inc/commonglpi.class.php:1125                      CommonGLPI->showPrimaryForm()
  front/ticket.form.php:181                          CommonGLPI->display()
2016-01-11 17:08:21 [27@adelaunay-ThinkPad-Edge-E320]
  *** PHP Notice(8): Undefined index: helpdesk_hardware
  Backtrace :
  inc/item_ticket.class.php:372                      
  inc/commonglpi.class.php:208                       Item_Ticket->getTabNameForItem()
  inc/ticket.class.php:663                           CommonGLPI->addStandardTab()
  inc/commonglpi.class.php:158                       Ticket->defineTabs()
  inc/commonglpi.class.php:602                       CommonGLPI->defineAllTabs()
  inc/commonglpi.class.php:1125                      CommonGLPI->showPrimaryForm()
  front/ticket.form.php:181                          CommonGLPI->display()
```

On computer (and ticket), i still have this behavior (no form displayed) : 

image


An additionnal question also, is a super-admin (or another profil with checkecd unlock right) can unlock object of another user ?

**Comment by tomolimo** [comment]
I still don't understand why you have such errors. I don't have any of these. 
Could it be possible that we are not using the same version of GLPI? 
I made this evolution based on the 0.91 master. Could it be the root cause? 

Regarding unlock rights: any user with a profile with UNLOCK rights for an item type can unlock any (regardless of other users) locked item of this type.

**Comment by tomolimo** [comment]
Which PHP version are you using?
May I have a copy of your DB + GLPI folder?

**Comment by tomolimo** [comment]
Did you change something in the 'Read-Only' profile? I mean: did you unchecked the READ rights for tickets?
The error messages let me think that the 'Read-Only' profile is not set or not defined in the database, could you check if in glpi_profiles the record 'Read-Only' profile is present and if the fields 'interface' and 'helpdesk_hardware' are filled in with something (should be 'central' and '0')?

If this profile is not set correctly it may come from the version which differ from mime (0.91): if yours is 0.90 then the upgrade function has not been executed or the initial DB has not been set correctly: this could lead to a profile not been set correctly.

**Comment by orthagh** [comment]
PHP 5.6.11-1ubuntu3.1 

> May I have a copy of your DB + GLPI folder?
It will be difficult to clean up.
i'm on your branch (uptodate) and with 0.90.2 migrated data, so i'm supposed to have same glpi folder 

on update test, i didn't change anythin in profile.
I just enabled lock for all possible itemtypes logout, login, and opening same object in different session (as we said previously)

Read only profile is here and seems ok.

I plan to continue tests this afternoon, i'll investigate deeply to find the diff

**Comment by orthagh** [comment]
on empty data, from 0.90/bugfixes to your branch, it's functionnal.

Some errors also at the time of update (not related to my problem): 

```
*** PHP Notice(8): Undefined index: lock_use_lock_item
  Backtrace :
  inc/toolbox.class.php:592                          
  inc/objectlock.class.php:441                       Toolbox::userErrorHandlerDebug()
  inc/commondbtm.class.php:3094                      ObjectLock::getSearchOptionsToAdd()
  inc/commondbtm.class.php:3220                      CommonDBTM->getSearchOptions()
  inc/commondbtm.class.php:3292                      CommonDBTM->getSearchOptionByField()
  inc/commondbtm.class.php:810                       CommonDBTM->filterValues()
  inc/profileright.class.php:258                     CommonDBTM->add()
  install/update_0901_091.php:170                    ProfileRight::updateProfileRights()
  install/update.php:791                             update0901to091()
  install/update.php:943                             updateDbUpTo031()
```

**Comment by tomolimo** [comment]
Yes I saw this also...
It's coming from that fact that during update, the getSearchOptionsToAdd function is called !!!
I really don't understand why it's called during update, I'll add a test on this array field.

**Comment by tomolimo** [comment]
committed a test to prevent these update error messages

**Comment by tomolimo** [comment]
Regarding error message of post (https://github.com/glpi-project/glpi/pull/363#issuecomment-170599016) we may also think that 'Read-Only' profile is not loaded from DB.
I'll check that and keep you informed

**Comment by tomolimo** [comment]
If I force in DB in table glpi_configs the field 'lock_lockprofile_id' to a non existing profiles_id value. then I get the same errors than you:
glpi - tickets - windows internet explorer - inprivate _2016-01-12_10-44-31
Could you check that this field is correctly set to the 'Read-Only' profile id? (for new installation should be 8)
thank you,
regards,
Tomolimo

**Comment by orthagh** [comment]
After update my data, i enabled lock in config and also choose manually this 'Read-Only' profile.
So yes, it's correctly set (and existing)

**Comment by tomolimo** [comment]
If you needed to set it: it means that it was not set during update? right?

**Comment by orthagh** [comment]
yup

**Comment by tomolimo** [comment]
then there is a problem during update as the update0901to091() is doing the correct set of this information into glpi_configs table.

    $query = "INSERT INTO `glpi_profiles` 
         (`name`, `interface`, `is_default`, `helpdesk_hardware`, `helpdesk_item_type`, `ticket_status`, `date_mod`, `comment`, `problem_status`, `create_ticket_on_login`, `tickettemplates_id`, `change_status`) 
         VALUES 
         ('Read-Only','central','0','0','[]','{\"1\":{\"2\":0,\"3\":0,\"4\":0,\"5\":0,\"6\":0},\"2\":{\"1\":0,\"3\":0,\"4\":0,\"5\":0,\"6\":0},\"3\":{\"1\":0,\"2\":0,\"4\":0,\"5\":0,\"6\":0},\"4\":{\"1\":0,\"2\":0,\"3\":0,\"5\":0,\"6\":0},\"5\":{\"1\":0,\"2\":0,\"3\":0,\"4\":0,\"6\":0},\"6\":{\"1\":0,\"2\":0,\"3\":0,\"4\":0,\"5\":0}}', NULL, 'This profile defines read-only access. It is used when objects are locked. It can also be used to give to users rights to unlock objects.', '{\"1\":{\"7\":0,\"2\":0,\"3\":0,\"4\":0,\"5\":0,\"8\":0,\"6\":0},\"7\":{\"1\":0,\"2\":0,\"3\":0,\"4\":0,\"5\":0,\"8\":0,\"6\":0},\"2\":{\"1\":0,\"7\":0,\"3\":0,\"4\":0,\"5\":0,\"8\":0,\"6\":0},\"3\":{\"1\":0,\"7\":0,\"2\":0,\"4\":0,\"5\":0,\"8\":0,\"6\":0},\"4\":{\"1\":0,\"7\":0,\"2\":0,\"3\":0,\"5\":0,\"8\":0,\"6\":0},\"5\":{\"1\":0,\"7\":0,\"2\":0,\"3\":0,\"4\":0,\"8\":0,\"6\":0},\"8\":{\"1\":0,\"7\":0,\"2\":0,\"3\":0,\"4\":0,\"5\":0,\"6\":0},\"6\":{\"1\":0,\"7\":0,\"2\":0,\"3\":0,\"4\":0,\"5\":0,\"8\":0}}', 0, 0, '{\"1\":{\"9\":0,\"10\":0,\"7\":0,\"4\":0,\"11\":0,\"12\":0,\"5\":0,\"8\":0,\"6\":0},\"9\":{\"1\":0,\"10\":0,\"7\":0,\"4\":0,\"11\":0,\"12\":0,\"5\":0,\"8\":0,\"6\":0},\"10\":{\"1\":0,\"9\":0,\"7\":0,\"4\":0,\"11\":0,\"12\":0,\"5\":0,\"8\":0,\"6\":0},\"7\":{\"1\":0,\"9\":0,\"10\":0,\"4\":0,\"11\":0,\"12\":0,\"5\":0,\"8\":0,\"6\":0},\"4\":{\"1\":0,\"9\":0,\"10\":0,\"7\":0,\"11\":0,\"12\":0,\"5\":0,\"8\":0,\"6\":0},\"11\":{\"1\":0,\"9\":0,\"10\":0,\"7\":0,\"4\":0,\"12\":0,\"5\":0,\"8\":0,\"6\":0},\"12\":{\"1\":0,\"9\":0,\"10\":0,\"7\":0,\"4\":0,\"11\":0,\"5\":0,\"8\":0,\"6\":0},\"5\":{\"1\":0,\"9\":0,\"10\":0,\"7\":0,\"4\":0,\"11\":0,\"12\":0,\"8\":0,\"6\":0},\"8\":{\"1\":0,\"9\":0,\"10\":0,\"7\":0,\"4\":0,\"11\":0,\"12\":0,\"5\":0,\"6\":0},\"6\":{\"1\":0,\"9\":0,\"10\":0,\"7\":0,\"4\":0,\"11\":0,\"12\":0,\"5\":0,\"8\":0}}'); 
         ";
    $DB->query( $query ) or die( "0.91 Can't add default Unlock profile" ) ;
    $ret = $DB->insert_id();

    // insert default rights for this profile
    ProfileRight::updateProfileRights($ret,
                  array( 'backup' => '1',
                           'bookmark_public' => '1',
                           'budget' => '161',
                           'calendar' => '1',
                           'cartridge' => '161',
                           'change' => '1185',
                           'changevalidation' => '0',
                           'computer' => '161',
                           'config' => '1',
                           'consumable' => '161',
                           'contact_enterprise' => '161',
                           'contract' => '161',
                           'device' => '0',
                           'document' => '161',
                           'domain' => '1',
                           'dropdown' => '1',
                           'entity' => '1185',
                           'followup' => '8193',
                           'global_validation' => '0',
                           'group' => '129',
                           'infocom' => '1',
                           'internet' => '129',
                           'itilcategory' => '1',
                           'knowbase' => '2177',
                           'knowbasecategory' => '1',
                           'link' => '129',
                           'location' => '1',
                           'logs' => '1',
                           'monitor' => '161',
                           'netpoint' => '1',
                           'networking' => '161',
                           'notification' => '1',
                           'password_update' => '0',
                           'peripheral' => '161',
                           'phone' => '161',
                           'planning' => '3073',
                           'printer' => '161',
                           'problem' => '1185',
                           'profile' => '129',
                           'project' => '1185',
                           'projecttask' => '1',
                           'queuedmail' => '1',
                           'reminder_public' => '129',
                           'reports' => '1',
                           'reservation' => '1',
                           'rssfeed_public' => '129',
                           'rule_dictionnary_dropdown' => '1',
                           'rule_dictionnary_printer' => '1',
                           'rule_dictionnary_software' => '1',
                           'rule_import' => '1',
                           'rule_ldap' => '1',
                           'rule_mailcollector' => '1',
                           'rule_softwarecategories' => '1',
                           'rule_ticket' => '1',
                           'search_config' => '0',
                           'show_group_hardware' => '1',
                           'sla' => '1',
                           'software' => '161',
                           'solutiontemplate' => '1',
                           'state' => '1',
                           'statistic' => '1',
                           'task' => '8193',
                           'taskcategory' => '1',
                           'ticket' => '7297',
                           'ticketcost' => '1',
                           'ticketrecurrent' => '1',
                           'tickettemplate' => '1',
                           'ticketvalidation' => '0',
                           'transfer' => '1',
                           'typedoc' => '1',
                           'user' => '2177') ) ;

    // updates rights for Super-Admin profile
    ProfileRight::updateProfileRights(4,
               array( 'budget' => '255',
                           'cartridge' => '255',
                           'change' => '1279',
                           'computer' => '255',
                           'consumable' => '255',
                           'contact_enterprise' => '255',
                           'contract' => '255',
                           'document' => '255',
                           'entity' => '3319',
                           'group' => '151',
                           'internet' => '159',
                           'knowbase' => '7319',
                           'link' => '151',
                           'monitor' => '255',
                           'networking' => '255',
                           'peripheral' => '255',
                           'phone' => '255',
                           'printer' => '255',
                           'problem' => '1279',
                           'profile' => '151',
                           'project' => '1279',
                           'reminder_public' => '151',
                           'rssfeed_public' => '151',
                           'software' => '255',
                           'ticket' => '128159',
                           'user' => '7327' ) ) ;


    Config::setConfigurationValues('core', array( 'lock_use_lock_item' => 0,
                                             'lock_autolock_mode'   => 1,
                                             'lock_directunlock_notification' => 0,
                                             'lock_item_list' => '[]',
                                             'lock_lockprofile_id' => $ret));

**Comment by orthagh** [comment]
if just found the error.

https://github.com/glpi-project/glpi/pull/363/files#diff-15cea08f657239af5265c8ad277bdd65R396
This part display a list of profiles with an array key => value
With the display_emptychoice params, my array 

```
Array
  (
      [1] => Client - Simple
      [2] => normal
      [3] => admin
      [4] => super-admin
      [7] => Client - Chef de projet
      [8] => Client - Superviseur
      [9] => Hotliner
      [11] => Supervisor
      [12] => TECLIB' Dispatch
      [13] => TECLIB' Helpdesk
      [14] => TECLIB' Experts
      [15] => TECLIB' utilisateurs internes
      [17] => Externe helpdesk
      [18] => test dico
      [19] => test super profile
      [32] => Copie de super-admin
      [33] => Copie de super-admin
      [34] => Copie de super-admin
      [35] => Stork MDM registered users
      [36] => Stork MDM guest users
      [37] => Read-Only
  )
```

changes to 

```
Array
  (
      [0] => -----
      [1] => Client - Simple
      [2] => normal
      [3] => admin
      [4] => super-admin
      [5] => Client - Chef de projet
      [6] => Client - Superviseur
      [7] => Hotliner
      [8] => Supervisor
      [9] => TECLIB' Dispatch
      [10] => TECLIB' Helpdesk
      [11] => TECLIB' Experts
      [12] => TECLIB' utilisateurs internes
      [13] => Externe helpdesk
      [14] => test dico
      [15] => test super profile
      [16] => Copie de super-admin
      [17] => Copie de super-admin
      [18] => Copie de super-admin
      [19] => Stork MDM registered users
      [20] => Stork MDM guest users
      [21] => Read-Only
  )
```

And so in CFG_GLPI[lock_lockprofile_id], the value 21 is set, so ...

the transformation is triggered here : 
https://github.com/glpi-project/glpi/blob/master/inc/dropdown.class.php#L1691

Array_unshift function seems to not consider this array as a associative (aka collection) one

i suggest using :

```php
Profile::dropdown(array('name'  => 'lock_lockprofile_id', 
                              'value' => $CFG_GLPI['lock_lockprofile_id']));
```

instead of  Dropdown::showFromArray
With this control, we have also include search input, and ajax loading.

**Comment by tomolimo** [comment]
You may find that not very funny, but I saw this bug, and made comments on it to Yllen in source code https://github.com/glpi-project/glpi/commit/7511b57d3b54329ec1d75b7e7a9dcf7f9de4f3e2#commitcomment-15090958

And yes, I fixed this bug on my local copy.
would you like me to commit this fix?

**Comment by tomolimo** [comment]
Any news?

**Comment by orthagh** [comment]
too busy today, i'll finish checks tomorrow

**Comment by tomolimo** [comment]
Hello,
Any news?
regards,
tomolimo

**Comment by orthagh** [comment]
Sorry for the delay (i'm very busy on some customers issues).
I need more tests to merged this pr.

But, for me, features is valid and accepted for 0.91, so the merge will coming soon

**Comment by orthagh** [comment]
I found some time to do tests : 
- OK for config, no more errors, and it's the good profile stored in db after save
- In displayLockMessage function, i prefer this position (it's very minor)

```javascript
position: {
   my: 'left top',
   at: 'left+20 top+30',
   of: $('#page'),
   collision: 'none'
},
```

- on the action "unlock item" for you current lock, we have 1 confirm (do you want to...) and 1 alert (item unlocked), I feel the second one is too much, your opinion ?

- we need to remove trailing spaces and do some coding standard (see https://forge.glpi-project.org/projects/glpi/wiki/CodingStandards, max lenght for line : 100) but we can do after merge


So no more blocking issues for me, we can merge (i'll do it friday)

**Comment by tomolimo** [comment]
For the position: with small monitors, you'll have such view:
image

'List' is missing and '< |<' is missing also

image

For the confirmation to reload: use-case is most of the time used to unlock items that have been forgotten during browser leaving. When this happens, the object is left locked in DB. So that when someone else wants to read-write the object he/she can't have the lock and then requests for unlock. If the reload of the page is automatic, then the object will be relocked by the same person such preventing the unlock requester to get the lock for himself/herself. 
So what do you prefer? leaving the page as it is (with locked info)? Or auto reloading the page (and such relocking it)?

For the coding norms: thank you, I'll be able to code like needed (I didn't know this page).

**Comment by orthagh** [comment]
* for position, maybe higher (under top menu, but before text).

* for 2nd confirm, is the use case of not reloading page exists ? do you have some users who unlock and don't want to relock item ?

In addition, I think jquery dialog (in modal) instead of native javascript confirms will be prettier

**Comment by tomolimo** [comment]
For position, do you mean here?
image

For the unlock/reload see the case I described above

**Comment by tomolimo** [comment]
for jquery dialog: I agree

**Comment by tomolimo** [comment]
I'll commit a new version tomorrow :)

**Comment by tomolimo** [comment]
I've got a problem: GLPi itself doesn't use the jquery dialog, but directly the  javascript confirm function:
image
Should I continue to switch to jquery dialog?

**Comment by orthagh** [comment]
Yes, please use the glpi function In Html class
I'll prepare a review of this function to use jquery dialog.

**Comment by tomolimo** [comment]
which function?

**Comment by orthagh** [comment]
Html::addConfirmationOnAction but after some read, it's not very usable...

I still search a way

**Comment by orthagh** [comment]
So, use window.confirm !

i push a PR for override theses natives dialog.
See here : https://github.com/glpi-project/glpi/pull/380

I would like your opinion on this pr

**Comment by tomolimo** [comment]
Nice looking! you may add a title to the dialog with option title: '$something' and make it optionnal

**Comment by orthagh** [comment]
currently working on this suggestion. Ty

**Comment by tomolimo** [comment]
Did you test notifications?
and cron?

**Comment by tomolimo** [comment]
Hello,
I changed a little bit history log management:
1) Prevented to log history when unlock is due to auto-lock / auto-unlock.
2) added history log when unlock is from massive-actions.
regards,
Tomolimo

**Comment by tomolimo** [comment]
Any news?

**Comment by orthagh** [comment]
In a big rush currently.

I don't forget you

> On Jan 22 2016, at 1:53 pm, tomolimo &lt;notifications@github.com&gt; wrote:  

>

> Any news?

>

> —  
Reply to this email directly or [view it on GitHub](https://github.com/glpi-pr
oject/glpi/pull/363#issuecomment-173912802).

**Comment by orthagh** [comment]
I just checked cron methods, they seem ok.
Maybe you can add a $task->log(...) with a link to the affected object (with $instance->getLink())

Also, instead of 60 * 60, you can use HOUR_TIMESTAMP ( I'm very picky :D ) 


I also checked notification code, it's ok for me.

Last thing, can you trail existing white spaces in you new classes ?

**Comment by tomolimo** [comment]
Hello,
Could you check and test this new commit?
1) If it fits you I'll change the dialogs all over GLPI to be compliant with the new ones.
2) I added a llink in the cron logs to unlocked objects
3) Added 'no-store' to 'cache-control' otherwise when doing a 'back' in browser the page is not reloaded and then object is not re-locked.

regards,
Tomolimo

**Comment by orthagh** [comment]
sure

**Comment by orthagh** [comment]
seems ok.
I think your new confirm and alert dialog would be in a better place in html.class.php (with native function redefinition). And with a more generic name (like confirmCallback and alertCallback for example).
So, we can re-use them in another place.

**Comment by tomolimo** [comment]
Yes I agree for the location and renaming,
But it will not be possible to use native redefinition for confirm: as jQuery.dialog is asynchronous and native confirm are synchronous.
For alert: it may be possible to use the native redefine.

**Comment by orthagh** [comment]
yes, i agree, my opinion was to have theses in addition of already defined confirm and alert

**Comment by tomolimo** [comment]
ok, then I'll put them into the html.class.php file, right now

**Comment by tomolimo** [comment]
need to be tested and merged, and then I'll create anew branch to try to adapt alert and confirm functions to the rest of GLPI

**Comment by orthagh** [comment]
Great !

I rebased (and squashed all commits) in one commit (ff247d9) and merged this commit in master.
I close this PR un-merged so.

I checked implementation and did some functional tests after merge, all seems ok.
I had some conflict on theses files : 

* inc/html.class.php
* install/mysql/glpi-0.91-empty.sql
* install/update_0901_091.php

But i think, i managed correctly the merge.
You could check to be sure.


And you can now work on your new branch.

**Comment by tomolimo** [comment]
Hello
I checked a fresh install on a new server: was OK!
I'm going to check a migration too, I'll keep you informed,
thank you

**Comment by tomolimo** [comment]
Hello,
Could you check this #422 ? 
Thank you,
Regards,
Tomolimo

**Comment by orthagh** [comment]
Merged

**Comment by orthagh** [comment]
I cleaned and review the updater function to 0.91 on your parts.
I inserted profilerights in sql because of a lot of session error (due to ProfileRight::updateProfileRights who need to do an add() in Log based on $_SESSION['glpiID']

Other changes are more for readability 

See : 1d8e4fd739cca87c5ee49971af27c8146e14c6de

**Comment by tomolimo** [comment]
OK, thank you

**Pull request closed without merging**